documentation build requires python2

Building the docs with Python 3.9.5 results in an error early on in nxdl_desc2rst.py:

>     raise "Something wrong with indentation on this line:\n" + line

This results in the following error:

>    TypeError: exceptions must derive from BaseException

It runs fine in python2, so I suspect that to be the issue. I might be wrong, though.
Would be nice to this migrate to supported python in the near future. 

Replacing the `raise` statement above with a `pass` builds successful with the newer python. So at least for the success path, the work required will be low.

With py27, the Python code runs but the conda environment (built with "Python<3" and the `.github/env-workflow.yml` file) needs a newer Sphinx version:

```bash
(py27) prjemian@zap:~/.../NeXus/definitions$ !make
make impatient-guide manual -C build
make: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build'
make html -C impatient-guide
make[1]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/impatient-guide'
sphinx-build -b html -d _build/doctrees   . _build/html
Running Sphinx v1.8.5
loading pickled environment... failed: unsupported pickle protocol: 5
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index                                                                                                                                                                          
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                                                                                                                                           
generating indices...
writing additional pages... search
copying images... [100%] NeXus-logo.png                                                                                                                                                                  
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

The HTML pages are in _build/html.
touch _build/html/.nojekyll

Build finished. The HTML pages are in _build/html.
make[1]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/impatient-guide'
make -C manual/source
make[1]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source'
make -C classes
make[2]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes'
make -C base_classes
make[3]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/base_classes'
echo "Adding summaries to base_classes/index.rst"
Adding summaries to base_classes/index.rst
python ../../../../utils/nxdl_summary.py base_classes
make[3]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/base_classes'
make -C applications
make[3]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/applications'
echo "Adding summaries to applications/index.rst"
Adding summaries to applications/index.rst
python ../../../../utils/nxdl_summary.py applications
make[3]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/applications'
make -C contributed_definitions
make[3]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/contributed_definitions'
echo "Adding summaries to contributed_definitions/index.rst"
Adding summaries to contributed_definitions/index.rst
python ../../../../utils/nxdl_summary.py contributed_definitions
make[3]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/contributed_definitions'
make[2]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes'
make[1]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source'
make html -C manual
make[1]: Entering directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual'
Running Sphinx v1.8.5

Sphinx version error:
This project needs at least Sphinx v2.3 and therefore cannot be built with this version.
make[1]: *** [Makefile:26: html] Error 2
make[1]: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build/manual'
make: *** [Makefile:18: manual] Error 2
make: Leaving directory '/home/prjemian/Documents/projects/NeXus/definitions/build'
```

The Sphinx version is not a problem for this issue.

--- a/.github/env-workflow.yml
+++ b/.github/env-workflow.yml
@@ -1,0 +1,12 @@
+name: nexusformat
+
+channels:
+  - defaults
+  - conda-forge
+  - aps-anl-tag
+
+dependencies:
+  - lxml
+  - pyRestTable
+  - Sphinx
+  - six

--- a/.github/workflows/syntax-checks.yml
+++ b/.github/workflows/syntax-checks.yml
@@ -40,8 +40,7 @@ jobs:
 
     - name: Install dependencies
       run: |
-        $CONDA/bin/conda env update --file environment.yml --name $ENV_NAME
-        # $CONDA/bin/conda install pytest --name $ENV_NAME
+        $CONDA/bin/conda env update --file .github/env-workflow.yml --name $ENV_NAME
         $CONDA/bin/conda list -r --name $ENV_NAME
       env:
         PY_VER: ${{ matrix.python-version }}

--- a/nxdl.xsd
+++ b/nxdl.xsd
@@ -1182,7 +1182,7 @@
 			<xs:documentation>
 				An ``enumeration`` restricts the values allowed for a specification.
 				Each value is specified using an ``item`` element, such as:
-				``<item value="Synchrotron X-ray Source" />``.
+				``&lt;item value="Synchrotron X-ray Source" />``.
 				Could contain these elements:
 				
 				* ``doc``

--- a/utils/nxdl_desc2rst.py
+++ b/utils/nxdl_desc2rst.py
@@ -10,6 +10,7 @@ the NXDL chapter.
 
 import os, sys
 import lxml.etree
+import textwrap
 
 
 TITLE_MARKERS = '- + ~ ^ * @'.split()  # used for underscoring section titles
@@ -353,39 +354,19 @@ def printDocs(ns, parent, indentLevel=0):
 
 
 def getDocFromNode(ns, node, retval=None):
-    docnodes = node.xpath('xs:annotation/xs:documentation', namespaces=ns)
-    if docnodes == None:
+    annotation_node = node.find('xs:annotation', ns)
+    if annotation_node is None:
         return retval
-    if not len(docnodes) == 1:
+    documentation_node = annotation_node.find('xs:documentation', ns)
+    if documentation_node is None:
         return retval
     
-    # be sure to grab _all_ content in the documentation
-    # it might look like XML
-    s = lxml.etree.tostring(docnodes[0], pretty_print=True)
-    p1 = s.decode().find('>')+1
-    p2 = s.decode().rfind('</')
-    text = s[p1:p2].decode().lstrip('\n')    # cut off the enclosing tag
-    
-    lines = text.splitlines()
-    if len(lines) > 1:
-        indent0 = len(lines[0]) - len(lines[0].lstrip())
-        indent1 = len(lines[1]) - len(lines[1].lstrip())
-        if len(lines) > 2:
-            indent2 = len(lines[2]) - len(lines[2].lstrip())
-        else:
-            indent2 = 0
-        if indent0 == 0:
-            indent = max(indent1, indent2)
-            text = lines[0]
-        else:
-            indent = indent0
-            text = lines[0][indent:]
-        for line in lines[1:]:
-            if not len(line[:indent].strip()) == 0:
-                raise IndentationError(
-                    "Something wrong with indentation on this line:\n" + line
-                )
-            text += '\n' + line[indent:]
+    # Be sure to grab _all_ content in the <xs:documentation> node.
+    # The content might look like XML.
+    s = lxml.etree.tostring(documentation_node, method="text", pretty_print=True)
+    rst = s.decode().lstrip('\n')  # remove any leading blank lines
+    rst = rst.rstrip()  # remove any trailing white space
+    text = textwrap.dedent(rst)  # remove common leading space
 
     # substitute HTML entities in markup: "<" for "&lt;"
     # thanks: http://stackoverflow.com/questions/2087370/decode-html-entities-in-python-string
@@ -483,7 +464,8 @@ if __name__ == '__main__':
     developermode = True
     developermode = False
     if developermode and len(sys.argv) != 2:
-        NXDL_SCHEMA_FILE = os.path.join('..', 'nxdl.xsd')
+        path = os.path.dirname(__file__)
+        NXDL_SCHEMA_FILE = os.path.join(path, '..', 'nxdl.xsd')
     else:
         if len(sys.argv) != 2:
             print("usage: %s nxdl.xsd" % sys.argv[0])

--- a/utils/nxdl_desc2rst.py
+++ b/utils/nxdl_desc2rst.py
@@ -209,7 +209,7 @@ DATATYPE_POSTAMBLE = '''
 def _tagMatch(ns, parent, match_list):
     '''match this tag to a list'''
     if parent is None:
-        raise "Must supply a valid parent node"
+        raise ValueError("Must supply a valid parent node")
     parent_tag = parent.tag
     tag_found = False
     for item in match_list:
@@ -335,7 +335,7 @@ def applyTemplates(ns, parent, path, indentLevel, handler=generalHandler):
             if name in ('nx:groupGroup',):
                 print(">"*45, name)
             if name in db:
-                raise "Duplicate name found: " + name
+                raise KeyError("Duplicate name found: " + name)
             db[name] = node
     for name in sorted(db):
         node = db[name]
@@ -382,7 +382,9 @@ def getDocFromNode(ns, node, retval=None):
             text = lines[0][indent:]
         for line in lines[1:]:
             if not len(line[:indent].strip()) == 0:
-                raise "Something wrong with indentation on this line:\n" + line
+                raise IndentationError(
+                    "Something wrong with indentation on this line:\n" + line
+                )
             text += '\n' + line[indent:]
 
     # substitute HTML entities in markup: "<" for "&lt;"

--- a/utils/nxdl_desc2rst.py
+++ b/utils/nxdl_desc2rst.py
@@ -362,7 +362,8 @@ def getDocFromNode(ns, node, retval=None):
         return retval
     
     # Be sure to grab _all_ content in the <xs:documentation> node.
-    # The content might look like XML.
+    # In the documentation nodes, use XML entities ("&lt;"" instead of "<")
+    # for documentation characters that would otherwise be considered as XML.
     s = lxml.etree.tostring(documentation_node, method="text", pretty_print=True)
     rst = s.decode().lstrip('\n')  # remove any leading blank lines
     rst = rst.rstrip()  # remove any trailing white space

